main/core.js

```diff
@@ -0,0 +1,171 @@
+  await start()
+}
+
+async function subscribeWithRetry (/** @type {Context} */ ctx) {
+  while (true) {
+    const controller = new AbortController()
+    try {
+      await subscribe(ctx, controller.signal)
+    } catch (err) {
+      controller.abort()
+      console.error(err)
+      await timers.setTimeout(1000)
+    }
+  }
+}
+
+async function subscribe (
+  /** @type {Context} */ ctx,
+  /** @type {AbortSignal} */ signal
+) {
+  await Promise.all([
+    (async () => {
+      for await (const metrics of core.metrics.follow(undefined, { signal })) {
+        ctx.setTotalJobsCompleted(metrics.totalJobsCompleted)
+      }
+    })(),
+    (async () => {
+      const it = core.activity.follow({ nLines: 0, signal })
+      for await (const activity of it) {
+        ctx.recordActivity(activity)
+        if (
+          activity.type === 'info' &&
+          activity.message.includes('Saturn Node is online')
+        ) {
+          online = true
+        } else if (
+          activity.message === 'Saturn Node started.' ||
+          activity.message.includes('was able to connect') ||
+          activity.message.includes('will try to connect')
+        ) {
+          online = false
+        }
+      }
+    })()
+  ])
+}
+
+async function start () {
+  assert(wallet.getAddress(), 'Core requires FIL address')
+  console.log('Starting Core...')
+
+  const childProcess = execa(corePath, {
+    env: {
+      FIL_WALLET_ADDRESS: wallet.getAddress(),
+      CACHE_ROOT: consts.CACHE_ROOT,
+      STATE_ROOT: consts.STATE_ROOT
+    }
+  })
+
+  /** @type {string | null} */
+  let exitReason = null
+
+  app.on('before-quit', () => {
+    childProcess.kill()
+  })
+
+  childProcess.on('close', code => {
+    console.log(`Core closed all stdio with code ${code ?? '<no code>'}`)
+
+    ;(async () => {
+      const log = await core.logs.get()
+      Sentry.captureException('Core exited', scope => {
+        // Sentry UI can't show the full 100 lines
+        scope.setExtra('logs', log.split('\n').slice(-10).join('\n'))
+        scope.setExtra('reason', exitReason)
+        return scope
+      })
+    })()
+  })
+
+  childProcess.on('exit', (code, signal) => {
+    const reason = signal ? `via signal ${signal}` : `with code: ${code}`
+    const msg = `Core exited ${reason}`
+    console.log(msg)
+    exitReason = signal || code ? reason : null
+  })
+}
+
+function isOnline () {
+  return online
+}
+
+async function maybeMigrateFiles () {
+  const oldSaturnDir = join(consts.LEGACY_CACHE_HOME, 'saturn')
+  const newSaturnDir = join(consts.CACHE_ROOT, 'modules', 'saturn-l2-node')
+  try {
+    await fs.stat(newSaturnDir)
+    return
+  } catch {}
+  try {
+    await fs.stat(oldSaturnDir)
+  } catch {
+    return
+  }
+  console.log(
+    'Migrating files from %s to %s',
+    oldSaturnDir,
+    newSaturnDir
+  )
+  await fs.mkdir(dirname(newSaturnDir), { recursive: true })
+  await fs.rename(oldSaturnDir, newSaturnDir)
+  console.log('Migration complete')
```

Comment:
Should we also delete the `LEGACY_CACHE_HOME` directory to avoid leaving no-longer-used stuff on the user's drive?

Reply:
We are moving `oldSaturnDir`, which is defined as follows:

```js
const oldSaturnDir = join(consts.LEGACY_CACHE_HOME, 'saturn')
```

I think `fs.rename` is going to move `LEGACY_CACHE_HOME/saturn`, but leave `LEGACY_CACHE_HOME` (the parent directory).

A potential problem is if `LEGACY_CACHE_HOME` is the same directory as the current `CACHE_ROOT`, in which case we should not delete it.

Considering the complexity, maybe this isn't worth the effort right now, and we can wait until somebody complains.

Reply:
Ah I see, I didn't get that this was about the parent directory. I'm also afraid of the negative side effects of changing this, let's revisit if someone complains